src/archive/src/main.rs

```diff
@@ -0,0 +1,469 @@
+    MEMORY_MANAGER.with(|cell| f(&*cell.borrow()))
+}
+
+/// A helper function to access the log.
+fn with_log<R>(f: impl FnOnce(&StableLog) -> R) -> R {
+    LOG.with(|cell| f(&*cell.borrow()))
+}
+
+/// A helper function to access the anchor-based index.
+fn with_anchor_index_mut<R>(f: impl FnOnce(&mut AnchorIndex) -> R) -> R {
+    ANCHOR_INDEX.with(|cell| f(&mut *cell.borrow_mut()))
+}
+
+/// Configuration state of the archive.
+enum ConfigState {
+    Uninitialized, // This state is only used between wasm module initialization and init().
+    Initialized(ArchiveConfig),
+}
+
+impl ConfigState {
+    fn get(&self) -> &ArchiveConfig {
+        match &self {
+            ConfigState::Uninitialized => trap("archive config not initialized"),
+            ConfigState::Initialized(config) => config,
+        }
+    }
+}
+
+/// Configuration of the archive canister.
+#[derive(Clone, Debug, CandidType, Deserialize)]
+struct ArchiveConfig {
+    /// This canister will accept entries only from this principal.
+    ii_canister: Principal,
+    /// The maximum number of entries returned in a single read canister call.
+    max_entries_per_call: u16,
+    /// Timestamp of the last install / upgrade of this canister.
+    last_upgrade_timestamp: Timestamp,
+}
+
+impl Storable for ConfigState {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        match &self {
+            ConfigState::Uninitialized => Cow::Borrowed(&[]),
+            ConfigState::Initialized(config) => {
+                let buf = candid::encode_one(config).expect("failed to encode archive config");
+                Cow::Owned(buf)
+            }
+        }
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        if bytes.len() == 0 {
+            return ConfigState::Uninitialized;
+        }
+        ConfigState::Initialized(
+            candid::decode_one::<ArchiveConfig>(&bytes).expect("failed to decode archive config"),
+        )
+    }
+}
+
+/// Index key for the anchor index.
+/// Changing the (serialized) size of this value requires a stable memory migration.
+#[derive(Debug)]
+struct AnchorIndexKey {
+    anchor: Anchor,
+    timestamp: Timestamp,
+    log_index: u64,
+}
+
+impl AnchorIndexKey {
+    /// Returns bytes corresponding only to the timestamp and log_index component of the anchor index key.
+    /// This is useful as an offset when doing a range scan with the anchor component.
+    fn to_anchor_offset(&self) -> Vec<u8> {
+        let mut buf =
+            Vec::with_capacity(std::mem::size_of::<Timestamp>() + std::mem::size_of::<u64>());
+        buf.extend(&self.timestamp.to_be_bytes());
+        buf.extend(&self.log_index.to_be_bytes());
+        buf
+    }
+}
+
+/// Storable implementation for the index key.
+/// Note: byte ordering is very important as the keys are sorted on a byte level (lower to higher index)
+/// --> use big endian to ensure that the most significant bytes are compared first
```

Comment:
If I understand correctly, the following entry would be encoded like this:
* anchor `10042` at time `123456`, at index `1`: `100421234561`

This way, by scanning/fetching indices lexicographically, we ensure that all anchor operations will be listed together, ordered first by timestamp, then by index. We can then deserialize the key and retrieve the index (last 8 bytes), used to index the log.

Is that correct?

* This wasn't trivial to figure out; how about adding a line or two about this?
* It looks like we rely on each field (anchor, timestamp) to have a known, fixed sized. What if the size changes (for instance changing the type of `Anchor`, do we fail silently? Does that invalidate the entire index?

Reply:
I expanded the documentation with this example. 👍